proposals/0010-stream-type.md

```diff
@@ -0,0 +1,162 @@
+- `"unknown"` - No media source is set, the source is not yet loaded (e.g. because `preload="none"`), or the stream type has yet to be determined.
+
+### Recommended computation for RFC8216 (aka HLS)
+
+`streamType = (#EXT-X-PLAYLIST-TYPE === "VOD") ? "on-demand" : "live"`
+
+__Context__:
+
+> A Media Playlist has further constraints on its updates if it contains an EXT-X-PLAYLIST-TYPE tag.  An EXT-X-PLAYLIST-TYPE tag with a value of VOD indicates that the Playlist file MUST NOT change.  An EXT-X-PLAYLIST-TYPE tag with a value of EVENT indicates that the Server MUST NOT change or remove any part of the Playlist file, with the exception of EXT-X-PART tags and Media Metadata tags as described above; the Server MAY append lines to the Playlist.
+
+  \- From: https://datatracker.ietf.org/doc/html/draft-pantos-hls-rfc8216bis-12#section-6.2.1
+
+#### NOTE: Why not `#EXT-X-ENDLIST` detection?
+
+The `#EXT-X-ENDLIST` tag only indicates that a particular playlist will no longer be updated with additional Media Segments or media segment parts (`#EXTINF` / `#EXT-X-PART-INF`). Using this value therefore conflates, e.g., an ended live stream with an on demand stream, and thus should not be relied upon for stream type detection.
+
+### Recommended computation for ISO/IEC 23009-1 (aka "MPEG-DASH")
+
+`streamType = (MPD@type === "dynamic") ? "live" : "on-demand`
+
+__Context__:
+
+MPD@type: 
+> default: `static`
+> specifies the type of the Media Presentation. For static Media Presentations (`@type="static"`), all Segments are available between the `@availabilityStartTime` and the `@availabilityEndTime`. For dynamic Media Presentations (`@type="dynamic"`), Segments typically have different availability times.
+
+\- From: *§5.3.1.2 Table 3 - Semantics of MPD element*
+
+### Recommended inferred value from native `HTMLMediaElement`
+
+```js
+streamType = Number.isFinite(mediaEl.duration)
+  ? "on-demand"
+  : mediaEl.duration === Number.POSITIVE_INFINITY
+  ? "live"
+  : "unknown"
+```
+
+__Context__:
+
+`media.duration`: 
+> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Returns the length of the media resource, in seconds, assuming that the start of the media resource is at time zero. <br/>
+> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Returns NaN if the duration isn't available. <br/>
+> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Returns Infinity for unbounded streams.
+
+\- From: https://html.spec.whatwg.org/multipage/media.html#dom-media-duration-dev
+
+
+## Event: `"streamtypechange"`
+
+This event should be dispatched from an extended `HTMLMediaElement` instance whenever the stream type has been determined for a loaded media source.
+
+__NOTE__: For media unload cases, implementors *__SHOULD__* ensure the `streamType` is (re)set to `"unknown"` after an unload begins but prior to any `"emptied"` external event handlers being invoked. This ensures that consumers of the API may reliably use the `"emptied"` event to monitor potential changes in `"streamType"`.
```

Comment:
I don't see a problem with this suggestion in general, but does this matter if the `streamtypechange` event exists? Feels like either we have good reasons for this event and we're telling devs they should listen for it when they care about live vs. vod UIs, or you can rely on existing events and we should spell out the events where stream type is likely to change. I'm not sure a mix is helpful?

Reply:
We can definitely make it a requirement to dispatch a `"streamtypechange"` even on unload. I agree that a mix isn't helpful, but it's the reality of other property state change events in `HTMLMediaElement`. I'd actually prefer always relying on `"streamtypechange"` if that's how you're leaning.

Reply:
> it's the reality of other property state change events in HTMLMediaElement

I don't doubt you're right, but which ones?

> if that's how you're leaning

If the event exists that's how I lean. It's smelly if streamType changes and the event doesn't fire. Unless maybe if I'm hearing you say the entire media element source state workflow already depends on certain _reset_ events like `emptied`.

Reply:
> I don't doubt you're right, but which ones?

A couple of examples:

1. `paused` state - If the src is already playing and then the src is unloaded (`mediaEl.src = null; mediaEl.load()`), the `paused` value will update from `false` to `true`, but there is not a corresponding event (other than `emptied`) to monitor for the change.
2. `duration` state - If the src is already loaded and then the source is unloaded (`mediaEl.src = null; mediaEl.load()`), the `duration` value will update from e.g. some finite numeric value to `NaN`, but there is not a corresponding event (other than `emptied`) to monitor for the change.

I'm pretty confident that there are more examples, but this provides a couple to indicate that "unload"/"emptied" cases are treated "on a different order", so to speak, with respect to media state events and state updates.

All that said, I'm still 💯 on board with dispatching `"streamtypechange"` for unload cases, since I personally think this is "the right way to do it" (aka if an event purports to model changes for some state, it should be dispatches _**whenever**_ that state changes).

Reply:
If this was implemented by browser vendors, would this event be necessary? Or would `loadedmetadata` or `durationchange` be sufficient?

Looking at the HTML spec, it specifically calls out that no extra events are dispatched as part of unloading. It'll either be an `emptied` event or an `error` event. You won't get any other events like `removetrack` or `durationchange` or `paused`. (see https://html.spec.whatwg.org/multipage/media.html#loading-the-media-resource:event-media-emptied-3)
There's a discussion on the spec around making a change for this https://github.com/whatwg/html/issues/4874
My first instinct would be to follow what the HTML spec does currently and update it when that issue is resolved.

Reply:
I was unaware of that discussion, good callout! What you're saying makes sense to me (tl;dr - expect to monitor `emptied` events for changes in the `streamtype` value because that's established in spec right now), in which case we would not change the proposal (unless/until there is an official resolution in the discussion referenced ☝️). @heff thoughts?

Reply:
> If this was implemented by browser vendors, would this event be necessary? Or would `loadedmetadata` or `durationchange` be sufficient?

@gkatsev If the browser venders supported native playback for the respective HAS specifications, I believe `loadedmetadata` would be both sufficient and appropriate (assuming they load + parse all necessary manifest/playlist files before dispatching, which I suspect they would).